### PR TITLE
dotnet add list remove operations

### LINKED ISSUE #1451: Document CRUD command for package references (dotnet add/remove/list package)

### LINKED ISSUE #1452: Document CRUD command for project references (dotnet add/remove/list project)

### LINKED ISSUE #1453: Document the base CRUD commands (add/list/remove)
The convention we have is to have `dotnet add/list/remove/update` verbs followed by a noun that "specifies"  what the verb will operate on. I think we should cover this convention somehow and explain it to people.

### PR DESCRIPTION
**Pull request by spboyer**
# dotnet CRUD operations

dotnet add remove list reference
dotnet add remove package

## Summary

RC4 Docs for CLI

Fixes #1451 
Fixes #1452 

## Suggested Reviewers

@mairaw @blackdwarf

### CODE REVIEW

File: docs/toc.md [hunk 1]
```diff
@@ -189,6 +189,14 @@
 ### [dotnet-nuget-delete](core/preview3/tools/dotnet-nuget-delete.md)
 ### [dotnet-nuget-push](core/preview3/tools/dotnet-nuget-push.md)
 ### [dotnet-nuget-locals](core/preview3/tools/dotnet-nuget-locals.md)
+### Project modification commands
```
Comment (mairaw) [inline docs/toc.md]: let's also add dotnet-sln to the toc since it's missing

File: docs/toc.md [hunk 2]
```diff
@@ -189,6 +189,14 @@
...
+#### References
+##### [dotnet-add-reference.md](core/preview3/tools/dotnet-add-reference.md)
```
Comment (mairaw) [inline docs/toc.md]: let's add these files to their final destination since preview3 location will cease to exist soon and so we avoid more redirections.

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ms.devlang: dotnet
+ms.assetid: 88e0da69-a5ea-46cc-8b46-5493242b7af9
+---
+# dotnet-add package
+
+[!INCLUDE[preview-warning](../../../includes/warning.md)]
+
+## Name
+
+`dotnet-add package` - Add package reference to a project file.
```
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: all of our description has an s in the end, so:
Add -> Adds a?

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+## Synopsis
+
+```
+dotnet add [<PROJECT>] package <PACKAGE_NAME> [-v|--version] [-f|--framework] [-s|--source] [--package-directory] [-h|--help]
```
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: should we remove help from this line?

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+## Description
+
+The `dotnet add package` command provides a convenient option to add a NuGet package reference to a project.
+
+## Arguments
+
+`PROJECT`
+
+Project name to use. If not specified, the command will search the current directory for one.
```
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: Project name to use. -> Project file to be modified.?

will search -> searches

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+`PACKAGE_NAME`
+
+Package reference to add.
```
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: Package -> The package?

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+## Options
+
+`-h|--help`
+
+Show help information for the command and/or argument
```
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: We usually write these descriptions using 3rd person. 

We probably should have an include file for the help command. 😄

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+`-v|--version`
+
+Version for the package to be added.
```
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: Version of the package?

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+`-f|--framework`
```
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: `-f|--framework <FRAMEWORK>`. etc.

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+Package reference to add.
+
+## Options
+
+`-h|--help`
+
+Show help information for the command and/or argument
+
+`-v|--version`
```
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: If you look at our other command pages, you'll see that I usually add the option argument too `-v|--version <PACKAGE_VERSION>`. Since we're not restricted in size, I don't use the exact string shown in --help

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+Version for the package to be added.
+
+`-f|--framework`
+
+Add reference only when targetting a specific framework
```
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: missing punctuation, 3rd person

reference -> package reference

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+Add reference without performing restore preview and compatibility check.
+
+`-s|--source`
+
+Use specific NuGet package sources to use during the restore.
+
+`--package-directory`
+
+Restore the packages to this Directory .
```
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: lowercase directory, remove extra space after Directory

**Review by mairaw — COMMENTED** [review]
Some comments before I leave today. I'll review the others later.

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+## Synopsis
+
+```
+dotnet add [<PROJECT>] package <PACKAGE_NAME> [-v|--version] [-f|--framework] [-s|--source] [--package-directory] [-h|--help]
+dotnet add package [-h|--help]
+```
+
+## Description
+
+The `dotnet add package` command provides a convenient option to add a NuGet package reference to a project.
```
Comment (blackdwarf) [inline docs/core/preview3/tools/dotnet-add-package.md]: We need to add more here. One thing to add would be that there is a compatibility check that happens. We should also state that the restore is ran after adding a package so that the asset file is complete. We should also add what is added to a project (a sample csproj fragment).

File: docs/core/preview3/tools/dotnet-add-package.md (hunk 7 above)
Comment (blackdwarf) [inline docs/core/preview3/tools/dotnet-add-package.md]: +1

File: docs/core/preview3/tools/dotnet-remove-reference.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+## Synopsis
+
+```
+dotnet remove [<PROJECT>] reference [-f|--framework] [args] [-h|--help]
+dotnet remove reference [-h|--help]
+```
+
+## Description
+
+The `dotnet remove reference` command provides a convenient option to remove a project references from a project.
```
Comment (blackdwarf) [inline docs/core/preview3/tools/dotnet-remove-reference.md]: As above, we should add more details here like the csproj fragment and links to MSBuild docs that define the `<ProjectReference>` element.

**Review by thesmithee — CHANGES_REQUESTED** [review]

File: docs/toc.md (hunk 1 above)
Comment (spboyer) [inline docs/toc.md]: done.

File: docs/toc.md (hunk 2 above)
Comment (spboyer) [inline docs/toc.md]: where is the "final desitination"? /preview4? 

There are a lot that need to move, wouldn't be easier to move them all when these are complete?

File: docs/core/preview3/tools/dotnet-add-package.md (hunk 3 above)
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-add-package.md]: done.

File: docs/core/preview3/tools/dotnet-add-package.md (hunk 6 above)
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-add-package.md]: done

File: docs/core/preview3/tools/dotnet-remove-reference.md (hunk 14 above)
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-remove-reference.md]: I added this to the dotnet-add-reference.md as it seems to make more sense there.

File: docs/core/preview3/tools/dotnet-add-package.md (hunk 13 above)
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-add-package.md]: Added this commentary

File: docs/core/preview3/tools/dotnet-remove-reference.md (hunk 14 above)
Comment (blackdwarf) [inline docs/core/preview3/tools/dotnet-remove-reference.md]: Makes sense. Maybe a sentence to that effect?

**Review by thesmithee — APPROVED** [review]
Some cosmetical changes but otherwise good to go. Let's merge this.

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+ms.devlang: dotnet
+ms.assetid: 88e0da69-a5ea-46cc-8b46-5493242b7af9
+---
+# dotnet-add package
+
+[!INCLUDE[preview-warning](../../../includes/warning.md)]
+
+## Name
+
+`dotnet-add package` - Adds a package reference to a project file. After running the command, there is a compatibility check to ensure the package attempting to be added is compatible with all of the frameworks in the project. If the check passes, a restore is run and the `<PackageReference>` fragment is added to the project file.
```
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-add-package.md]: done.

**Comment by mairaw** [comment]
I'll take a look at this today.

File: docs/core/preview3/tools/dotnet-add-reference.md [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+## Synopsis
+
+```
+dotnet add [<PROJECT>] reference [-f|--framework] [args] [-h|--help]
+dotnet add reference [-h|--help]
+```
+
+## Description
+
+The `dotnet add reference` command provides a convenient option to add project references to a project. After running the command, the `<ProjectReference>` fragments are added to the project file.
```
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-add-reference.md]: Done.

File: docs/core/preview3/tools/dotnet-add-reference.md [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+Add a project reference:
+
+`dotnet add app/app.csproj reference lib/lib.csproj`
+
+Add multiple project references:
+
+`dotnet add reference lib1/lib1.csproj lib2/lib2.csproj`
+
+Add multiple project references using globbing pattern:
```
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-add-reference.md]: 👍

**Comment by mairaw** [comment]
I see we need some edits. I'll work on them before merging. Thanks!

**Comment by mairaw** [comment]
Let me know if you guys are cool with my last changes @spboyer @blackdwarf. Thanks!

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 18]
```diff
@@ -17,20 +17,20 @@ ms.assetid: 88e0da69-a5ea-46cc-8b46-5493242b7af9
...
 ## Name
 
-`dotnet-add package`
+`dotnet-add package` - Adds a package reference to a project file.
 
 ## Synopsis
 
 ```
-dotnet add [<PROJECT>] package <PACKAGE_NAME> [-v|--version] [-f|--framework] [-s|--source] [--package-directory] [-h|--help]
+dotnet add [<PROJECT>] package <PACKAGE_NAME> [-v|--version] [-f|--framework] [-s|--source] [--package-directory]
```
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-add-package.md]: Removing `[-h|--help]` from this line infers that you cannot use the option with `package` or `dotnet add` which you can. I suggest leaving this.

File: docs/core/preview3/tools/dotnet-add-package.md [hunk 19]
```diff
@@ -69,38 +69,38 @@ The package reference to add.
 
 `-h|--help`
 
-Prints out a short help for the command or argument.
+Prints out a short help for the command.
```
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-add-package.md]: again you can use `-h` for the argument `add` and/or `package` which displays different or additional information relative to that argument.

File: docs/core/preview3/tools/dotnet-add-reference.md [hunk 20]
```diff
@@ -17,12 +17,12 @@ ms.assetid: 5e2a3efd-443c-4f23-a1b1-a662a5387879
...
 ## Name
 
-`dotnet-add reference` - Add project to project references.
+`dotnet-add reference` - Adds project to project references.
 
 ## Synopsis
 
 ```
-dotnet add [<PROJECT>] reference [-f|--framework] [args] [-h|--help]
+dotnet add [<PROJECT>] reference [-f|--framework] <PROJECT_REFERENCES>
```
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-add-reference.md]: same comment here about removing `-h`

File: docs/core/preview3/tools/dotnet-list-reference.md [hunk 21]
```diff
@@ -17,33 +17,37 @@ ms.assetid: 8f954a0c-03f8-4fbc-a529-b313ab12c623
...
 ## Name
 
-`dotnet-list reference` - Remove project to project references.
+`dotnet-list reference` - Lists project to project references.
 
 ## Synopsis
 
 ```
-dotnet list [<PROJECT>] reference [-h|--help]
+dotnet list [<PROJECT>] reference
```
Comment (spboyer) [inline docs/core/preview3/tools/dotnet-list-reference.md]: same comment about removing `-h`

**Review by spboyer — COMMENTED** [review]
LGTM other than the feedback concerning removal of `-h|--help`.

We also might consider linking to the [globbing](http://www.tldp.org/LDP/abs/html/globbingref.html) patterns as a + if you think its important.

File: docs/core/preview3/tools/dotnet-add-package.md (hunk 18 above)
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: I know it's possible, but would you recommend people to use --help with any other option? The result will always be the same as `dotnet add package -h`. Also, the first line is only good for saying you can do `dotnet add package packagename [other options] --help`, but it doesn't cover the `dotnet add -h` case, which is the parent command.

File: docs/core/preview3/tools/dotnet-add-package.md (hunk 19 above)
Comment (mairaw) [inline docs/core/preview3/tools/dotnet-add-package.md]: Sure, but this page is about dotnet-add-package, not dotnet-add in general.

Comment (spboyer) [inline docs/core/preview3/tools/dotnet-add-package.md]: perhaps it is inferred that `-h` is always available regardless. However, there is not a **dotnet-add.md** and IMO the documentation should reflect it.

Let's not hold this up.

**Comment by mairaw** [comment]
I like the idea of the link for the globbing pattern. I just find that one hard to follow. Prefer the [Wikipedia](https://en.wikipedia.org/wiki/Glob_(programming)) or http://man7.org/linux/man-pages/man7/glob.7.html. I'll create an issue for this and do this in a global scope, since we mention glob in other parts too. Let's merge this! Thanks @spboyer.
